### ISSUE TITLE
Rule proposal: no-null (revisited)

### ISSUE THREAD

**Issue by octogonz** [eslint/eslint#12177]
Let's reopen the discussion https://github.com/eslint/eslint/issues/6701 regarding adding a `no-null` rule as part of ESLint core.

Currently this is solved by [eslint-plugin-no-null](https://github.com/nene/eslint-plugin-no-null), however it's awkward to rely on an external NPM package for such a simple and basic rule.

The previous issue was closed with the opinion:

> To me this seems like a very personal preference, so I'm 👎 on inclusion in the core. This would be better as a custom rule.

I'd like to offer some counterpoints:

- ESLint [announced](https://eslint.org/blog/2019/01/future-typescript-eslint) that it would be replacing TSLint, which provided [no-null-keyword](https://palantir.github.io/tslint/rules/no-null-keyword/) as a core rule.  Thus, this is a gap for people migrating from TSLint.
- When I worked at Microsoft, hundreds of projects used TSLint to forbid `null`
- Now I work at HBO, and they apply this rule in a massive internal monorepo

Thus, this practice is at least somewhat mainstream.  Why is that?  (The next section is optional reading heheh.)

## Why forbid `null` instead of `undefined`?

Most programming languages have a "null" or "nil" value that functions...

1. as the initial value for an uninitialized variable; and
2. as the value of `x.y` or `x["y"]` when x has no such key; and
3. as a special token that developers can assign to indicate an unknown or empty state.

In JavaScript, the `undefined` value fulfills roles (1), (2) and (3).  JavaScript's `null` value is a redundant secondary token that can only be used for (3), even though its name confusingly implies otherwise.  From this perspective, the `null` value was arguably a mistake in the original JavaScript language design.  Nearly all other programming languages get along just fine with a single "null"/"nil" concept.  In actual practice, JavaScript loses no expressiveness if we always write `undefined` instead of `null`.  

One potential hitch is external libraries such as `JSON.parse("null")` and some popular NPM packages that return `null` from their APIs.  However, if we relax the lint rule to allow `null` comparisons, these cases rarely produce lint errors.

By contrast, banning `undefined` values would be impossible, since `null` cannot satisfy the roles (1) and (2).

Of course, it's completely valid to use `null` and `undefined` together.  But this practice requires developers to keep two concepts straight in their heads.  In TypeScript, it sometimes requires clunky type annotations like `string | null | undefined`.  Thus while using `null` is valid, and perhaps more historically idiomatic, it adds complexity without providing any significant value.

**Comment by octogonz** [comment]
@platinumazure right, I'm aware that it can be solved using `no-restricted-syntax`.  Here's what I'm currently using:

```json
  rules: [
          {
            "selector": ":not(BinaryExpression:matches([operator='!=='], [operator='==='])) > Literal[value=null]",
            "message": "Usage of \"null\" is deprecated except when received from legacy APIs; use \"undefined\" instead"
          }
  ]
```

`no-restricted-syntax` is fine for workarounds.  But it works by creating an array of different customized restrictions that all get the generic label `no-restricted-syntax` when reporting errors, or suppressing them.  This isn't a good general solution.  Core rules should have a specific name that we can search for with Google, and refer to in discussions.  (Otherwise, let's open a PR to delete all ESLint rules that can be simulated using `no-restricted-syntax` heheh.)

If banning `null` was an obscure policy that one isolated team wants, then `no-restricted-syntax` would be pretty reasonable.  However I think I've shown evidence that it's is a widespread practice.

**Comment by ilyavolodin** [comment]
> ESLint announced that it would be replacing TSLint, which provided no-null-keyword as a core rule. Thus, this is a gap for people migrating from TSLint.
Just a clarification, ESLint is not looking to replace TSLint. TSLint was deprecated by the team that was developing it. typescript-eslint parser was selected as a suggested replacement (it's a separate project, that doesn't have much to do with ESLint itself). So any typescript specific rules, or rules that are requested for feature parity with TSLint should go into typescript-eslint project, not ESLint core itself.
If you step away from the typescript issue, and look at this rule as more of a JavaScript generic rule, I still agree with the previous position, of this is a personal preference of the developer, and as such more suited to plugin than to the core.

**Comment by octogonz** [comment]
> ESLint is not looking to replace TSLint. TSLint was deprecated by the team that was developing it. typescript-eslint parser was selected as a suggested replacement (it's a separate project, that doesn't have much to do with ESLint itself)

Here's Palantir's announcement, which details how they met in person with the TypeScript owners and made a strategic decision to join forces with ESLint, with the goal "to avoid bifurcating the linting tool space" between JavaScript and TypeScript: https://medium.com/palantir/tslint-in-2019-1a144c2317a9

And here's the announcement from the ESLint blog, expressing their enthusiasm and support for linting TypeScript:  https://eslint.org/blog/2019/01/future-typescript-eslint

> So any typescript specific rules, or rules that are requested for feature parity with TSLint should go into typescript-eslint project, not ESLint core itself.

The **typescript-eslint** project generally doesn't accept rules unless they are  TypeScript-specific.  Even if the rule needs special logic for TypeScript, they prefer to implement it in ESLint and then extend from it, so that both JavaScript and TypeScript can benefit.

> If you step away from the typescript issue, and look at this rule as more of a JavaScript generic rule, I still agree with the previous position, of this is a personal preference of the developer, and as such more suited to plugin than to the core.

My company has a very large legacy code base that is pure JavaScript, and `null` has always been avoided there as well.

Anyway, there's the evidence that this is a somewhat common practice at major corporations.  If you still disagree and are convinced that it's merely an obscure "personal preference", could you provide some evidence to back that up?  I'm open-minded about it.

**Comment by octogonz** [comment]
One more data point (from @nene's original issue):  The [coding guidelines](https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines#null-and-undefined) for the TypeScript compiler project say:

**Comment by octogonz** [comment]
One more data point (from @nene's original issue):  The [coding guidelines](https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines#null-and-undefined) for the TypeScript compiler project say, simpy:  

> ### `null` and `undefined`
> 1. Use **undefined**. Do not use null.

They aren't advising everyone to do that, but there's a major code base, made by developers who certainly understand JavaScript pretty well.

**Comment by nzakas** [comment]
Hi everyone - A while ago the ESLint team made the decision that we would no longer be adding new core rules that simply disallowed some syntax. The primary reason for this is we knew if we continued down that road then everyone would want a rule to specifically disallow their least favorite syntax and we'd end up with hundreds of rules just for that. When we heard feedback that people didn't want to have to manually create and maintain their own plugins for just one rule, we added the `no-restricted-syntax` rule, which would allow anyone to define syntax they wanted to disallow.

Discussions about whether or not `null` or `undefined` is more preferred in any particular segment of the JavaScript community don't really weigh into the decision here. Given that we have nearly 300 rules already, the bar for adding any new rule is particularly high, and this proposal doesn't qualify.

Just to review, if you want a rule to disallow `null`, you do have several options:

1. Use the `eslint-plugin-no-null` package
1. Use `no-restricted-syntax` to get the same behavior with an existing core rule
1. Create a custom rule for your use (and perhaps distribute it through a plugin)

Given that the first two options don't require any work, they both seem like good approaches.

Thanks for the suggestion. We do appreciate thoughtful proposals like this one even though we aren't always able to accommodate.
